01 fib

jonathanheaden opened the pull request:
Fixes #221

#### Review of colleague's PR #164 

#### Changes proposed in this PR:
Implement Fib func
Add tests for fib(7) and fib(-7)

jonathanheaden committed:
Implement fib func

Initial draft

camscale commented on 01_fib/jonathanheaden/main.go:
```diff
@@ -0,0 +1,34 @@
+)
+
+var out io.Writer = os.Stdout
+var fibs = map[int]int{}
```
a map where each index is a sequential number from 0 is really the same data structure as an array/slice, just less efficient.

Also do you really need to store *all* the fibs, or just the last two? Can you get away with out an indexed data structure at all?

jonathanheaden replied:
agree. I have replaced the map with a three element array

camh-anz commented on 01_fib/jonathanheaden/main.go:
```diff
@@ -0,0 +1,34 @@
+	fibs[0] = 1
+	fibs[1] = 1
+	for i := 0; i < n; i++ {
+		if i > 1 {
```
Rather than have this if statement, can you just start the loop at 2?

camh-anz replied:
maybe not, since you need to print the earlier numbers...

camscale commented on 01_fib/jonathanheaden/main_test.go:
```diff
@@ -0,0 +1,24 @@
+	main()
+
+	// Then
+	expected := strconv.Quote("\n1\n1\n2\n3\n5\n8\n13\n\n-1\n-1\n-2\n-3\n-5\n-8\n-13\n")
```
The output of the negative fib series is not correct. Double-check the wikipedia page. The sign alternates.

jonathanheaden replied:
I had to think through this a lot, I see what you mean. I have altered the test and code

juliaogris reviewed (dismissed):
Please have a read through the [collected lab 1 review comments](https://github.com/anz-bank/go-course/wiki/Lab-1:-Collected-Code-Review-Comments):
- Create issue and add link to PR description
- Fix up markdown in Pr
- Consider using table driven tests
- Fold your 8 commits into one or two with `git rebase`

Your solution could be simplified to something like:

```go
	sign := // 1 for positive n ; -1 for negative n
	n1, n2 := 0, 1
	for j := 0; j < n*sign; j++ {
		n1, n2 = n2, n1+sign*n2
		// print n1
	}
```

Commit:
Implement Fib and add tests

Fib uses two vars to store the fib numbers as they are incremented. Tests cover positive, negative and fib(0)

juliaogris commented on 01_fib/jonathanheaden/main.go:
```diff
@@ -0,0 +1,39 @@
+		if prefix = ""; neg && (i%2 == 0) {
+			prefix = "-"
+		}
+		switch i {
```
No need to have a switch on loop index inside look - might as well do it before imo.

camscale reviewed (approved):
I happy with the Go code in this PR. You could implement the loop more simply along the lines of Julia's earlier suggestion, and a `fib(1)` test case would be nice to see as well, since you have a `case 1` in the code, but these are more software engineering issues and we want to focus on Go coding for these lab PRs.

juliaogris reviewed (approved):
The lab is solved!

You have done everything that's required.
I still think when you calculate fib you care about the right values as integer and not just them being printed correctly, so the prefix is a bid odd to me.
To me a solution like the listing below would have been a bit cleaner:
```go
func fib(n int) {
	if n == 0 {
		return
	}

	sign := 1
	if n < 0 {
		sign = -1
	}

	n1, n2 := 0, 1
	for j := 0; j < n*sign; j++ {
		n1, n2 = n2, n1+sign*n2
		fmt.Fprintln(out, n1)
	}
}
```

jonathanheaden committed:
Merge branch 'master' into 01-fib
